File: prez/code/saxpy_cublas.c
#include <stdlib.h>
#include <math.h>
#include <cublas.h>
#define N 8192

int main()
{

  int i;
  float alpha = 2.0f;
  float * x[N], y[N];
  float * xd, *yd;

  cublasInit();
  cublasSetVector(N,  sizeof(x[0]), x, 1, xd, 1);
  cublasSetVector(N,  sizeof(y[0]), y, 1, yd, 1);

  cublasSaxpy(N, alpha, xd, 1, yd, 1);
  
  cublasGetVector(N,  sizeof(y[0]), yd, 1, y, 1);
  cublasShutdown();

  for(i = 0; i < N; i ++ )
    if (fabsf(y[i] - 4.0) > 1e-6) exit (-1);
  
  return 0;
}

File: prez/code/saxpy_cublas.f90
program saxpy_cublas
  use cublas
  use iso_c_binding
  implicit none
  integer, parameter :: N = 8192
  type(cublasHandle) :: handle
  real(c_float), allocatable :: x(:) 
  real(c_float), allocatable :: y(:)   
  real(c_float), device :: xd(N), yd(N)
  real(c_float), device :: alpha = 2.0
  integer :: ierr, elemsize
  allocate(x(N), y(N))

  x = 1.0
  y = 2.0
  elemsize = sizeof(x(1))
  ierr = cublasCreate(handle)
  ierr = cublasSetVector(N, elemsize, x, 1, xd, 1)
  ierr = cublasSetVector(N, elemsize, y, 1, yd, 1)
  call cublasSaxpy(N, 2.0, xd, 1, yd, 1)
  ierr = cublasGetVector(N, elemsize, yd, 1, y, 1)
  ierr = cublasDestroy(handle)
  if (any(y /= 4.0)) stop -1

  deallocate(x, y)

end program saxpy_cublas
